pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml

```diff
@@ -0,0 +1,343 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data
+        xmlns="http://pmd.sourceforge.net/rule-tests"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
+    <test-code>
+        <description>violation: reassigned 'for' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>5</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i = 5; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: incremented 'for' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>5</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i++; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: incremented other variable</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            x++;
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: incremented field</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            this.i++; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: add to 'for' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>5</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i += 1; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>no violation: incremented 'for' loop variable</description>
+        <rule-property name="forReassign">skip</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i++;
```

Comment:
I don't see a real reason for allowing to skip these. In this case and nearly all other test cases, the increment is executed at the end of *every* iteration, so it would be clearer to remove it and replace the for update with `i += 2`.

Wrapping the increment inside another control statement (like an `if`) should not trigger the rule though, e.g.
```java
        for (int i=0; i < 10; i++) {
           doSomethingWith(i);
           if (something()) i++; // skip one sometimes
        }
```
I'm not sure we can do that precisely before having some reliable local control flow analysis. E.g. the following should also not trigger the rule:
```java
        for (int i=0; i < 10; i++) {
           doSomethingWith(i);
           if (something()) continue; 
           i++;
        }
```
I think, we'd better keep it like it is until we can do better

Reply:
I've looked into that and I think it's actually possible to do some limited kind of control flow analysis:

I can check if either
- the skip happens in some kind of control statement or 
- the skip is after a continue statement.

IMO, there are only a few cases where plain control flow analysis could work (e.g. if the skip is executed in every branch of an if/else or switch/case).
Apart from that, control statements are either constant conditions which can/should be covered by other rules or use variables or external logic and can't be analyzed statically.

I updated the tests accordingly. So far it seems  to be working!

Reply:
I've changed the logic to make sure that changing the variable in the expression of a control statement is also reported. So ´if(foo(i++)) doSomething();´ triggers the rule, while ´if (something()) i++;´ does not.

Reply:
Very well done. We can perfect that incrementally based on future false-positive/negative reports, but for now this is a good starting point